[Frontend] Load standard libarary in CompilerInstance::setup

ahoppen opened the pull request:
Instead of checking that the stdlib can be loaded in a variety of places, check it when setting up the compiler instance. This required a couple more checks to avoid loading the stdlib in cases where it’s not needed.

To be able to differentiate stdlib loading failures from other setup errors, make `CompilerInstance::setup` return an error message on failure via an inout parameter. Consume that error on the call side, replacing a previous, more generic error message, adding error handling where appropriate or ignoring the error message, depending on the context.

--- 

Was discussed in https://github.com/apple/swift/pull/39631#discussion_r739175831

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/Frontend/Frontend.cpp:
```diff
@@ -1065,6 +1079,13 @@ void CompilerInstance::performSema() {
 }
 
 bool CompilerInstance::loadStdlibIfNeeded() {
+  if (!FrontendOptions::doesActionRequireSwiftStandardLibrary(
+          Invocation.getFrontendOptions().RequestedAction)) {
+    return false;
+  }
+  if (Invocation.getFrontendOptions().InputsAndOutputs.shouldTreatAsLLVM()) {
+    return false;
+  }
```
Can these checks be sunk down into `CompilerInvocation::getImplicitStdlibKind`?

ahoppen replied:
I think it would make sense, but I’m currently hitting issues when scanning dependencies because apparently dependency scanning requires the stdlib to be loaded but returns fails in `FrontendOptions::doesActionRequireSwiftStandardLibrary` (see other comment).

hamishknight replied:
I actually realize now `doesActionRequireSwiftStandardLibrary` shouldn't be sunk into `getImplicitStdlibKind`, see my other comment

hamishknight commented on lib/DependencyScan/DependencyScanningTool.cpp:
```diff
@@ -162,7 +162,10 @@ DependencyScanningTool::initCompilerInstanceForScan(
 
   // Setup the instance
-  Instance->setup(Invocation);
+  std::string InstanceSetupError;
+  if (Instance->setup(Invocation, InstanceSetupError)) {
+    return std::make_error_code(std::errc::not_supported);
+  }
```
@artemcm Mind double checking it's correct we now bail here?

artemcm replied:
Yeah, that looks correct to me.

hamishknight commented on lib/Frontend/ModuleInterfaceBuilder.cpp:
```diff
@@ -191,13 +191,6 @@ bool ModuleInterfaceBuilder::buildSwiftModuleInternal(
     InputInfo.getPrimarySpecificPaths().SupplementaryOutputs;
     StringRef OutPath = OutputInfo.ModuleOutputPath;
 
-    // Bail out if we're going to use the standard library but can't load it. If
-    // we don't do this before we try to build the interface, we could end up
-    // trying to rebuild a broken standard library dozens of times due to
-    // multiple calls to `ASTContext::getStdlibModule()`.
-    if (SubInstance.loadStdlibIfNeeded())
-      return std::make_error_code(std::errc::not_supported);
-
```
@nkcsgexi @artemcm Mind double checking that it's okay this check is now done for all clients of `InterfaceSubContextDelegateImpl::runInSubCompilerInstance`? In particular, it *appears* that dependency scanning uses it, and `ModuleInterfaceLoaderOptions` seems to be currently configured with actions that would require a stdlib to be loaded, so if I'm reading it right, this may now cause the stdlib to be loaded when scanning dependencies for a swiftinterface. Is that correct?

ahoppen replied:
Actually, I believe it’s the other way round. Previously, we were always loading the stdlib for dependency scanning https://github.com/apple/swift/blob/0bd5ddb73acb88d72ff04ac88cb36e34dfd04b05/test/ScanDependencies/module_deps_cache_reuse.swift#L51
but dependency scanning isn’t listed as requiring a stdlib to be loaded https://github.com/apple/swift/blob/0bd5ddb73acb88d72ff04ac88cb36e34dfd04b05/lib/Frontend/FrontendOptions.cpp#L99
so it might not be loaded anymore after this change. Should we change `ScanDependencies` to return `true` in `doesActionRequireSwiftStandardLibrary` and thus also make it return `false` for `shouldActionOnlyParse`?

ahoppen replied:
Actually, I believe it’s the other way round. Previously, we were always loading the stdlib for dependency scanning https://github.com/apple/swift/blob/0bd5ddb73acb88d72ff04ac88cb36e34dfd04b05/test/ScanDependencies/module_deps_cache_reuse.swift#L51
but dependency scanning isn’t listed as requiring a stdlib to be loaded https://github.com/apple/swift/blob/0bd5ddb73acb88d72ff04ac88cb36e34dfd04b05/lib/Frontend/FrontendOptions.cpp#L99
so it might not be loaded anymore after this change. Should we change `ScanDependencies` to return `true` in `doesActionRequireSwiftStandardLibrary` and thus also make it return `false` for `shouldActionOnlyParse`?

hamishknight replied:
@ahoppen As I understand it, dependency scanning shouldn't need to load modules, it should still be able to pick up the implicit stdlib import though. I actually realize now this means that `doesActionRequireSwiftStandardLibrary` *shouldn't* be sunk into `getImplicitStdlibKind`, as when scanning dependencies, we need to know whether an implicit stdlib import is available, but we shouldn't actually load it.

So I believe `doesActionRequireSwiftStandardLibrary` & `shouldActionOnlyParse` are correct, but the logic that kicks of dependency scanning needs to be updated to not use the `Typecheck` or `EmitModuleOnly` frontend actions via `ModuleInterfaceLoaderOptions`, e.g:

https://github.com/apple/swift/blob/b02dc65e911e7aa4aff519ebaef7f092fedb86dd/lib/DependencyScan/ScanDependencies.cpp#L1381

Or `runInSubCompilerInstance` needs to otherwise be changed to make sure not to load the stdlib for dep scanning.

hamishknight commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1253,17 +1253,6 @@ static bool performCompile(CompilerInstance &Instance,
   if (opts.InputsAndOutputs.shouldTreatAsLLVM())
     return compileLLVMIR(Instance);
 
-  // If we aren't in a parse-only context and expect an implicit stdlib import,
-  // load in the standard library. If we either fail to find it or encounter an
-  // error while loading it, bail early. Continuing the compilation will at best
-  // trigger a bunch of other errors due to the stdlib being missing, or at
-  // worst crash downstream as many call sites don't currently handle a missing
-  // stdlib.
```
IMO it would be nice to preserve some of the rationale in this comment for the `loadStdlibIfNeeded` call in `setup`, what do you think?

ahoppen replied:
I’m adjusting it slightly and keeping it in.

hamishknight reviewed (approved):
Looks great, thank you! I'm not too familiar with a couple of the places this affects though, so I've tagged a couple of folks to take a look

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on include/swift/Frontend/ModuleInterfaceLoader.h:
```diff
@@ -311,6 +311,9 @@ struct ModuleInterfaceLoaderOptions {
     case FrontendOptions::ActionType::TypecheckModuleFromInterface:
       requestedAction = FrontendOptions::ActionType::Typecheck;
       break;
+    case FrontendOptions::ActionType::ScanDependencies:
+      requestedAction = Opts.RequestedAction;
+      break;
```
@nkcsgexi @artemcm Mind double checking that there aren't other callers of `CompilerInstance::setup` where we may need to change the requested action to avoid loading the stdlib? In particular I'm not too sure what arguments are being fed for the other dependency scanning call-sites, are they guaranteed to use the dependency scanning action?

artemcm replied:
This looks fine to me.
I also fully expect all dependency scanning call-sites to also have `FrontendOptions::ActionType::ScanDependencies`.

hamishknight commented on tools/SourceKit/lib/SwiftLang/SwiftDocSupport.cpp:
```diff
@@ -1467,8 +1474,7 @@ SourceFile *SwiftLangSupport::getSyntacticSourceFile(
 
-  if (ParseCI.setup(Invocation)) {
-    Error = "Compiler invocation set up failed";
+  if (ParseCI.setup(Invocation, Error)) {
```
Just noticed this case appears to only need the compiler instance for syntactic queries, but uses `initCompilerInvocationNoInputs` which will set the requested action to `Typecheck` and therefore I *believe* will now load the stdlib where it didn't previously. Should we overwrite the requested action before calling `setup`, or perhaps plumb through a parameter on `initCompilerInvocationNoInputs` to determine whether we want a semantic action?

ahoppen replied:
Ah, good catch. I added a commit to explicitly specify the frontend action when setting up the compiler invocation. This surfaced another case where we don’t need to load the stdlib either.

ahoppen committed:
[SourceKit] Explicitly specify frontend action type when creating compiler invocation

Explicitly specify the frontend action to make sure we aren’t loading thes stdlib if the performed action is syntactic only.

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[Frontend] Load standard libarary in CompilerInstance::setup

Instead of checking that the stdlib can be loaded in a variety of places, check it when setting up the compiler instance. This required a couple more checks to avoid loading the stdlib in cases where it’s not needed.

To be able to differentiate stdlib loading failures from other setup errors, make `CompilerInstance::setup` return an error message on failure via an inout parameter. Consume that error on the call side, replacing a previous, more generic error message, adding error handling where appropriate or ignoring the error message, depending on the context.

ahoppen commented:
@swift-ci Please smoke test
